### DISCUSSION TITLE
Use prosemirror schema instead of node extension API?

### DISCUSSION THREAD

**Discussion by tslocke** [ueberdosis/tiptap#2077] [General Questions] [answered]
Hi

Subject says it - is there any way to pass in a regular ProseMirror schema when creating a tiptap editor, as opposed to using `Node.create` and passing in extensions?

Thanks

**Comment by philippkuehn** [comment, marked as answer]
TLDR: Not really.

You can use the `onBeforeCreate` event to overwrite the schema. At this time the view is not created.

```js
import { Editor } from '@tiptap/core'

const editor = new Editor({
  onBeforeCreate() {
    this.editor.schema = new Schema({
      // ...
    })
  },
})
```

But I think this will throw a ProseMirror error because tiptap creates the schema right before `onBeforeCreate`. Without a `topNode` and a `text` extension, ProseMirror throws an error.

You can use a minimal setup to create a valid schema. After that you can overwrite it.

```js
import { Editor } from '@tiptap/core'
import Document from '@tiptap/extension-document'
import Paragraph from '@tiptap/extension-paragraph'
import Text from '@tiptap/extension-text'

const editor = new Editor({
  extensions: [
    Document,
    Paragraph,
    Text,
  ],
  onBeforeCreate() {
    this.editor.schema = new Schema({
      // ...
    })
  },
})
```

But that just doesn’t feel right.

I already thought about an option to pass a custom schema but you’ll miss some features that tiptap provides. Such as

- `keepOnSplit` option for attributes and marks, which is used for our custom `splitBlock` and `splitListItem` commands
- adding global attributes. A feature which is used by many extensions (`UniqueID`, `Color`, `FontFamily`, `TextAlign`) which would be not compatible
- probably some more features I’m not think of now

To summarize: it’s probably not a good idea if you want to use tiptap. 

Just out of interest: why do you want to put pass a schema?

**Reply by tslocke** [reply]
Thanks for the great reply.

As to the why, I am doing a lot of deep customisation in my app. I find it much easier to learn and think about customising ProseMirror, than having to think about two layers. For example where I need a plugin I try to write a ProseMirror plugin, rather than a TipTap extension. TipTap has been great as a quick way to get started but in the long run I may migrate to ProseMirror directly.

At the same time, I might be missing out on more than I realise if I go that way. The list of features in your reply is useful for that. Is there somewhere where the features that are unique to TipTap and not present in ProseMirror are documented?

Thanks again.

**Reply by tslocke** [reply]
BTW I think the `onBeforeCreate` above should be:

```
onBeforeCreate({editor}) {
  editor.schema = ...
}
```

**Reply by philippkuehn** [reply]
both should work
